src/api.py

```diff
@@ -0,0 +1,239 @@
+from StringIO import StringIO
+import traceback
+import os, json, uuid
+from os.path import join
+import conf, utils, adaptor
+from flex.core import validate
+import flex
+import connexion
+import flask
+from flask import request
+import main as scraper, validate as ajson_validate
+import jsonschema
+from connexion.resolver import RestyResolver
+from werkzeug.exceptions import HTTPException
+import logging
+
+LOG = logging.getLogger(__name__)
+
+#
+# utils
+#
+
+def validate_schema():
+    "validates the api scheme"
+    path = join(conf.PROJECT_DIR, 'schema', 'api.yaml')
+    spec = flex.load(path)
+    validate(spec)
+    return True
+
+def http_exception(code, msg):
+    "returns an HTTPException object with the correct code and message set"
+    ex = HTTPException(msg)
+    ex.code = code
+    return ex
+
+def http_ensure(case, msg, code=400):
+    "if not case, raise a client error exception (by default)"
+    if not case:
+        raise http_exception(code, msg)
+
+def listfiles(path, ext_list=None):
+    "returns a pair of (basename_list, absolute_path_list) for given dir, optionally filtered by extension"
+    path_list = map(lambda fname: os.path.abspath(join(path, fname)), os.listdir(path))
+    if ext_list:
+        path_list = filter(lambda path: os.path.splitext(path)[1] in ext_list, path_list)
+    path_list = sorted(filter(os.path.isfile, path_list))
+    return map(os.path.basename, path_list), path_list
+
+#
+#
+#
+
+def upload_folder():
+    return flask.current_app.config['UPLOAD_FOLDER']
+
+class BotLaxResolver(RestyResolver):
+    """custom resolver that flattens func name lookup after initial module.
+    RestyResolver parent assumes icky classes with .post and .search methods"""
+
+    def resolve_operation_id_using_rest_semantics(self, operation):
+        # ll: api.xml.search
+        orig_path = super(BotLaxResolver, self).resolve_operation_id_using_rest_semantics(operation)
+        bits = orig_path.split('.') # ll ['api', 'xml', 'search']
+
+        module = bits[0] # ll: 'api'
+        # the funname resolution is wonky, so discard it
+        method = bits[-1] # ll: 'search'
+
+        path = operation.path # ll: /article-json/validation/{filename}
+        # we want something like:
+        # /article-json/validation/{filename} => article_json_validation
```

Comment:
it's not an old comment, it's just illustrating what happens when a path contains path parameters and gets flattened. the final path has the method (`get`, `post`, etc) prefixed to it so that should help to differentiate, but if our future selves experience a conflict in function name resolution then we're probably doing it wrong.